include/deal.II/lac/linear_operator.h

```diff
@@ -1367,6 +1473,125 @@ linear_operator(const LinearOperator<Range, Domain, Payload> &operator_exemplar,
+  typename = typename std::enable_if<
+    !std::is_lvalue_reference<OperatorExemplar>::value>::type>
+LinearOperator<Range, Domain, Payload>
+linear_operator(OperatorExemplar &&, const Matrix &)
+{
+  Assert(false,
+         ExcMessage(
+           "You are trying to construct a linear operator with a temporary "
+           "operator_exemplar object."));
+}
+
+template <
+  typename Range   = Vector<double>,
+  typename Domain  = Range,
+  typename Payload = internal::LinearOperatorImplementation::EmptyPayload,
+  typename OperatorExemplar,
+  typename Matrix,
+  typename =
+    typename std::enable_if<!std::is_lvalue_reference<Matrix>::value>::type,
+  typename = typename std::enable_if<
+    !std::is_lvalue_reference<OperatorExemplar>::value>::type>
+LinearOperator<Range, Domain, Payload>
+linear_operator(OperatorExemplar &&, Matrix &&)
+{
+  Assert(false,
+         ExcMessage(
+           "You are trying to construct a linear operator from a temporary "
+           "matrix object."));
+}
+
+
+template <
+  typename Range   = Vector<double>,
+  typename Domain  = Range,
+  typename Payload = internal::LinearOperatorImplementation::EmptyPayload,
+  typename Matrix,
+  typename =
+    typename std::enable_if<!std::is_lvalue_reference<Matrix>::value>::type>
+LinearOperator<Range, Domain, Payload>
+linear_operator(const LinearOperator<Range, Domain, Payload> &, Matrix &&)
+{
+  Assert(false,
+         ExcMessage(
+           "You are trying to construct a linear operator from a temporary "
+           "matrix object."));
+}
+
+template <
+  typename Range   = Vector<double>,
+  typename Domain  = Range,
+  typename Payload = internal::LinearOperatorImplementation::EmptyPayload,
+  typename Matrix,
+  typename =
+    typename std::enable_if<!std::is_lvalue_reference<Matrix>::value>::type>
+LinearOperator<Range, Domain, Payload>
+linear_operator(Matrix &&)
+{
+  Assert(false,
+         ExcMessage(
+           "You are trying to construct a linear operator from a temporary "
+           "matrix object."));
+}
+
+template <typename Payload,
+          typename Solver,
+          typename Preconditioner,
+          typename Range  = typename Solver::vector_type,
+          typename Domain = Range,
+          typename        = typename std::enable_if<
+            !std::is_lvalue_reference<Preconditioner>::value>::type,
+          typename = typename std::enable_if<
+            !std::is_same<Preconditioner, PreconditionIdentity>::value>::type,
+          typename = typename std::enable_if<
+            !std::is_same<Preconditioner,
+                          LinearOperator<Range, Domain, Payload>>::value>::type>
+LinearOperator<Domain, Range, Payload>
+inverse_operator(const LinearOperator<Range, Domain, Payload> &,
+                 Solver &,
+                 Preconditioner &&)
+{
+  Assert(false,
+         ExcMessage(
+           "You are trying to construct an inverse operator with a temporary "
+           "preconditioner object."));
+}
```

Comment:
@bangerth This turns out to be the only way to get an error message of some sorts... (Note that you cannot simply use a `static_assert` in this context because SFINAE only applies to template argument substitution and not the body of the function.) Further, all of these `enable_if`s are really necessary because this function signature wins compared to the variants above.